typing: Convert module to TypeScript.

LaPulgaaa opened the pull request:
<!-- Describe your pull request here.-->
 -> First commit adds `server_typing_started_wait_period_milliseconds`, `server_typing_stopped_wait_period_milliseconds` property to `page_params.ts`.

-> Second commit extends `Recipient` type and corresponding changes in `typing_status.js.flow` ,`typing_status.test.js` (to satisfy the tests).

-> Third commit converts module - `typing.js` to `TypeScript`.

discussion:[CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/typing-js)

**Screenshots and screen captures:**

https://github.com/zulip/zulip/assets/123243429/882f42e9-738e-4264-b2e4-a411588244f9

https://github.com/zulip/zulip/assets/123243429/06fe547d-9ab9-4964-bb24-558403a5ae00

LaPulgaaa commented on web/src/typing.ts:
```diff
@@ -47,16 +55,16 @@ function send_stream_typing_notification(stream_id, topic, operation) {
     if (message_type === "direct" && user_settings.send_private_typing_notifications) {
         send_direct_message_typing_notification(to, operation);
-    } else if (message_type === "stream" && user_settings.send_stream_typing_notifications) {
+    } else if ("stream_id" in to && user_settings.send_stream_typing_notifications) {
```
TypeScript can not figure out whether `to` contains `stream_id` property by mere `message_type==="stream"` check.

timabbott replied:
Let's keep the check readable as originally written -- checking the message type.

I think a `!` type assertion or an explicit assert would be better for informing TypeScript that this means there's a `stream_id` field in the object ... though probably even better would be to improve the typing notifications types so it knows already.

timabbott replied:
Hmm I guess the best fix would be to do a prep commit extending the `Recipient` type in `typing_status.ts` to always be an object with an explicit `message_type` field, and be a union of two types, one with `message_type=stream` and the other with `message_type=direct`.

LaPulgaaa replied:
Extending `Recipient` to always be an object effectively means to re-type `typing_status.ts` as a lot of function leverage the fact of `Recipient` being` number[]` type sometimes. To maintain check readability  I think we can create a new type with disciminated union with structure same as of your proposed `Recipeint` type ie. an object with an explicit `message_type` property ……

LaPulgaaa replied:
I got no problem with mutating `Recipient` but I think it would be good if we can come up with something else for a problem which is local to this function block only.
Would go with your thoughts ......

chdinesh1089 replied:
I agree that extending `Recipient` type is a good refactor in general but `typing_status.ts` is also being used the mobile app. The changes on the web app won't be complex but I don't remember the details on how this would work for the mobile app. If it's a lot of hassle co-ordinating this with the mobile app, I'd lean towards fixing it here itself

amanagr replied:
Maybe it makes sense for `StreamTopic` to be a class so that we use `to instanceof StreamTopic` and keep this function clean?

LaPulgaaa commented on web/src/typing.ts:
```diff
@@ -16,11 +18,10 @@ import {user_settings} from "./user_settings";
+    void channel.post({
         url: "/json/typing",
         data,
-        success() {},
```
Empty success callback removed to avoid [eslint-error](https://eslint.org/docs/latest/rules/no-empty-function).

timabbott commented on web/src/typing.ts:
```diff
@@ -16,11 +18,10 @@ import {user_settings} from "./user_settings";
 
-function send_typing_notification_ajax(data) {
-    channel.post({
+function send_typing_notification_ajax(data: string | JQuery.PlainObject | undefined): void {
```
Do the callers pass these other types? This seems very likely wrong.

LaPulgaaa replied:
Agreed ....a rather loose check ..will fixed this.

LaPulgaaa commented:
Updated the PR with a prep commit , updated the PR discription with an explanation of the changes , added two screen records of local testing. Ready for your feedbacks @timabbott .

LaPulgaaa commented:
I have addressed the review comments , changed the PR discription to add video capture of manual local testing . The CI was failing because of large cache so I rebased it after the fix. This is ready for another review @timabbott

LaPulgaaa commented on web/tests/typing_status.test.js:
```diff
@@ -50,12 +50,12 @@ run_test("basics", ({override, override_rewire}) => {
 
     function notify_server_start(recipient) {
-        assert.deepStrictEqual(recipient, [1, 2]);
+        assert.deepStrictEqual(recipient, {message_type: "direct", ids: [1, 2]});
```
For direct message `recipient ` is not an array anymore. So tests have to changed a little.

timabbott commented:
Cool, thanks for doing that refactor. I reviewed this, noted a further cleanup we could do (pushed as a bonus commit), and reformatted the commit messages to be line-wrapped (as is the standard Git style). With those changes, this looks good, marked to merge once CI passes. Thanks @LaPulgaaa!

LaPulgaaa commented:
Thanks !! @timabbott . Don't know why the puppeteers are failing . I have seen same weird behavior this a few other PRs today.
